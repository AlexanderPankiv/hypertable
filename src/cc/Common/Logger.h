/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
#ifndef HYPERTABLE_LOGGER_H
#define HYPERTABLE_LOGGER_H

#include <iostream>

#include <log4cpp/Category.hh>
#include <log4cpp/Priority.hh>

#define DUMP_CORE *((int *)0) = 1;

namespace Hypertable {
  class Logger {
  public:
    static log4cpp::Category *logger;
    static void initialize(const char *name, log4cpp::Priority::Value priority=log4cpp::Priority::DEBUG);
    static void set_level(log4cpp::Priority::Value priority);
    static void set_test_mode(const char *name);
    static void suppress_line_numbers();
    static bool ms_show_line_numbers;
  };
}

#ifndef DISABLE_LOG_ALL

#ifndef DISABLE_LOG_DEBUG

#define HT_LOG_ENTER \
  if (Logger::logger->isDebugEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->debug("(%s:%d) %s() ENTER", __FILE__, __LINE__, __func__); \
    else \
      Logger::logger->debug("%s() ENTER", __func__); \
    std::cout << std::flush; \
  } while(0)

#define HT_LOG_EXIT \
  if (Logger::logger->isDebugEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->debug("(%s:%d) %s() EXIT", __FILE__, __LINE__, __func__); \
    else \
      Logger::logger->debug("%s() EXIT", __func__); \
    std::cout << std::flush; \
  } while(0)

#define HT_DEBUG(msg) if (Logger::logger->isDebugEnabled()) do { \
  Logger::logger->debug("(%s:%d) " msg, __FILE__, __LINE__) \
} while (0)

#define HT_DEBUGF(msg, ...) \
  if (Logger::logger->isDebugEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->debug("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->debug(msg, __VA_ARGS__);  \
    std::cout << std::flush; \
  } while (0)
#else
#define HT_LOG_ENTER
#define HT_LOG_EXIT
#define HT_DEBUG(msg)
#define HT_DEBUGF(msg, ...)
#endif

#ifndef DISABLE_LOG_INFO
#define HT_INFO(msg) if (Logger::logger->isInfoEnabled()) do { \
  Logger::logger->info("(%s:%d) " msg, __FILE__, __LINE__); \
} while(0)

#define HT_INFOF(msg, ...) \
  if (Logger::logger->isInfoEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->info("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->info(msg, __VA_ARGS__);  \
  } while (0)
#else
#define HT_INFO(msg)
#define HT_INFOF(msg, ...)
#endif

#ifndef DISABLE_LOG_NOTICE
#define HT_NOTICE(msg) if (Logger::logger->isNoticeEnabled()) do { \
  Logger::logger->notice("(%s:%d) " msg, __FILE__, __LINE__); \
} while (0)
#define HT_NOTICEF(msg, ...) \
  if (Logger::logger->isNoticeEnabled()) { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->notice("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->notice(msg, __VA_ARGS__);  \
  } while (0)
#else
#define HT_NOTICE(msg)
#define HT_NOTICEF(msg, ...)
#endif

#ifndef DISABLE_LOG_WARN
#define HT_WARN(msg) if (Logger::logger->isWarnEnabled()) do { \
  Logger::logger->warn("(%s:%d) " msg, __FILE__, __LINE__); \
} while (0)

#define HT_WARNF(msg, ...) \
  if (Logger::logger->isWarnEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->warn("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->warn(msg, __VA_ARGS__);  \
    std::cout << std::flush; \
   } while (0)
#else
#define HT_WARN(msg)
#define HT_WARNF(msg, ...)
#endif

#ifndef DISABLE_LOG_ERROR
#define HT_ERROR(msg) if (Logger::logger->isErrorEnabled()) do { \
  Logger::logger->error("(%s:%d) " msg, __FILE__, __LINE__); \
} while (0)
#define HT_ERRORF(msg, ...) \
  if (Logger::logger->isErrorEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->error("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->error(msg, __VA_ARGS__);  \
    std::cout << std::flush; \
  } while (0)
#else
#define HT_ERROR(msg)
#define HT_ERRORF(msg, ...)
#endif

#ifndef DISABLE_LOG_CRIT
#define HT_CRIT(msg) if (Logger::logger->isCritEnabled()) do { \
  Logger::logger->crit("(%s:%d) " msg, __FILE__, __LINE__); \
} while (0)
#define HT_CRITF(msg, ...) \
  if (Logger::logger->isCritEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->crit("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->crit(msg, __VA_ARGS__);  \
    std::cout << std::flush; \
  } while (0)
#else
#define HT_CRIT(msg)
#define HT_CRITF(msg, ...)
#endif

#ifndef DISABLE_LOG_ALERT
#define HT_ALERT(msg) if (Logger::logger->isAlertEnabled()) do { \
  Logger::logger->alert("(%s:%d) " msg, __FILE__, __LINE__); \
} while (0)
#define HT_ALERTF(msg, ...) \
  if (Logger::logger->isAlertEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->alert("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->alert(msg, __VA_ARGS__);  \
    std::cout << std::flush; \
  } while (0)
#else
#define HT_ALERT(msg)
#define HT_ALERTF(msg, ...)
#endif

#ifndef DISABLE_LOG_EMERG
#define HT_EMERG(msg) if (Logger::logger->isEmergEnabled()) do { \
  Logger::logger->emerg("(%s:%d) " msg, __FILE__, __LINE__); \
} while (0)
#define HT_EMERGF(msg, ...) \
  if (Logger::logger->isEmergEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->emerg("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->emerg(msg, __VA_ARGS__);  \
    std::cout << std::flush; \
  } while (0)
#else
#define HT_EMERG(msg)
#define HT_EMERGF(msg, ...)
#endif

#ifndef DISABLE_LOG_FATAL
#define HT_FATAL(msg) if (Logger::logger->isFatalEnabled()) do { \
  Logger::logger->fatal("(%s:%d) " msg, __FILE__, __LINE__); \
  abort(); \
} while (0)
#define HT_FATALF(msg, ...) \
  if (Logger::logger->isFatalEnabled()) do { \
    if (Logger::ms_show_line_numbers) \
      Logger::logger->fatal("(%s:%d) " msg, __FILE__, __LINE__, __VA_ARGS__);  \
    else \
      Logger::logger->fatal(msg, __VA_ARGS__);  \
    std::cout << std::flush; \
    abort(); \
  } while (0)
#define HT_EXPECT(_e_, _code_) if (_e_); else \
  HT_FATAL("failed expectation: " #_e_)
#else
#define HT_FATAL(msg)
#define HT_FATALF(msg, ...)
#define HT_EXPECT(_e_, _code_) if (_e_); else throw \
  Exception(_code_, "failed expectation: " #_e_)
#endif

#else // HYPERTABLE_DISABLE_LOGGING

#define HT_DEBUG(msg)
#define HT_DEBUGF(msg, ...) 
#define HT_INFO(msg)
#define HT_INFOF(msg, ...)
#define HT_NOTICE(msg)
#define HT_NOTICEF(msg, ...)
#define HT_WARN(msg)
#define HT_WARNF(msg, ...)
#define HT_ERROR(msg)
#define HT_ERRORF(msg, ...)
#define HT_CRIT(msg)
#define HT_CRITF(msg, ...)
#define HT_ALERT(msg)
#define HT_ALERTF(msg, ...)
#define HT_EMERG(msg)
#define HT_EMERGF(msg, ...)
#define HT_FATAL(msg)
#define HT_FATALF(msg, ...)
#define HT_EXPECT(e)
#define HT_LOG_ENTER
#define HT_LOG_EXIT

#endif // HYPERTABLE_DISABLE_LOGGING


#endif // HYPERTABLE_LOGGER_H
