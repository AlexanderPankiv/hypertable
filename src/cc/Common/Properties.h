/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_PROPERTIES_H
#define HYPERTABLE_PROPERTIES_H

#include <boost/program_options.hpp>

#include "Common/ReferenceCount.h"
#include "Common/Error.h"

// Required declarations for custom validators (otherwise no specializations)
namespace boost { namespace program_options {

typedef std::vector<std::string> Strings;
typedef std::vector<int64_t> Int64s;

extern void validate(boost::any &v, const Strings &values, int64_t *, int);
extern void validate(boost::any &v, const Strings &values, int32_t *, int);
extern void validate(boost::any &v, const Strings &values, uint16_t *, int);

// pre 1.35 vector<T> validate doesn't pickup user defined validate for T
#if BOOST_VERSION < 103500
template<typename T>
void validate(boost::any& v, const Strings &s, std::vector<T>*, int) {
  if (v.empty())
      v = boost::any(std::vector<T>());

  std::vector<T>* tv = boost::any_cast< std::vector<T> >(&v);
  assert(NULL != tv);

  for (unsigned i = 0; i < s.size(); ++i) {
    try {
      // so we can pick up user defined validate for T here
      boost::any a;
      Strings sv;
      sv.push_back(s[i]);
      validate(a, sv, (T*)0, 0);
      tv->push_back(boost::any_cast<T>(a));
    }
    catch(const bad_lexical_cast& /*e*/) {
      boost::throw_exception(invalid_option_value(s[i]));
    }
  }
}
#endif // < boost 1.35

}} // namespace boost::program_options

// convenience/abbreviated accessors
#define HT_PROPERTIES_ABBR_ACCESSORS \
  inline bool get_bool(const String &name) { return get<bool>(name); } \
  inline String get_str(const String &name) { return get<String>(name); } \
  inline Strings get_strs(const String &name) { return get<Strings>(name); } \
  inline uint16_t get_i16(const String &name) { return get<uint16_t>(name); } \
  inline int32_t get_i32(const String &name) { return get<int32_t>(name); } \
  inline int64_t get_i64(const String &name) { return get<int64_t>(name); } \
  inline Int64s get_i64s(const String &name) { return get<Int64s>(name); } \
  inline bool get_bool(const String &name, bool default_value) { \
    return get(name, default_value); \
  } \
  inline String get_str(const String &name, const String &default_value) { \
    return get(name, default_value); \
  } \
  inline Strings get_strs(const String &name, const Strings &default_value) { \
    return get(name, default_value); \
  } \
  inline uint16_t get_i16(const String &name, uint16_t default_value) { \
    return get(name, default_value); \
  } \
  inline int32_t get_i32(const String &name, int32_t default_value) { \
    return get(name, default_value); \
  } \
  inline int64_t get_i64(const String &name, int64_t default_value) { \
    return get(name, default_value); \
  } \
  inline Int64s get_i64s(const String &name, const Int64s &default_value) { \
    return get(name, default_value); \
  }

namespace Hypertable {

namespace Po = boost::program_options;

typedef std::vector<String> Strings;
typedef std::vector<int64_t> Int64s;

namespace Property {

const uint64_t K = 1000;
const uint64_t KiB = 1024;
const uint64_t M = K * 1000;
const uint64_t MiB = KiB * 1024;
const uint64_t G = M * 1000;
const uint64_t GiB = MiB * 1024;

// Abbrs for some common types
inline Po::typed_value<bool> *boo() { return Po::value<bool>(); }
inline Po::typed_value<String> *str() { return Po::value<String>(); }
inline Po::typed_value<Strings> *strs() { return Po::value<Strings>(); }
inline Po::typed_value<uint16_t> *i16() { return Po::value<uint16_t>(); }
inline Po::typed_value<int32_t> *i32() { return Po::value<int32_t>(); }
inline Po::typed_value<int64_t> *i64() { return Po::value<int64_t>(); }
inline Po::typed_value<Int64s> *i64s() { return Po::value<Int64s>(); }

} // namespace Property

typedef Po::options_description PropertiesDesc;
typedef Po::positional_options_description PositionalDesc;

class Properties : public ReferenceCount {
  typedef Po::variable_value Value;
  typedef Po::variables_map Map;
  typedef std::map<String, String> AliasMap;

public:
  Properties() : m_need_alias_sync(false) {}
  Properties(const String &filename, const PropertiesDesc &desc,
             bool allow_unregistered = false)
    : m_need_alias_sync(false) { load(filename, desc, allow_unregistered); }

  /**
   * load a property config file
   *
   * @param filename - name of the config file
   * @param desc - properties description
   * @param allow_unregistered - allow unregistered properties
   */
  void load(const String &filename, const PropertiesDesc &desc,
            bool allow_unregistered = false);

  /**
   * Parse arguments. Throws CONFIG_INVALID_ARGUMENT on error
   *
   * @param argc - argument count (from main)
   * @param argv - argument array (from main)
   * @param desc - options description
   * @param hidden - hidden options description
   * @param p - positional options description
   */
  void
  parse_args(int argc, char *argv[], const PropertiesDesc &desc,
             const PropertiesDesc *hidden = 0, const PositionalDesc *p = 0);

  /**
   * Same as above, except taking vector of strings as arguments
   */
  void
  parse_args(const std::vector<String> &args, const PropertiesDesc &desc,
             const PropertiesDesc *hidden = 0, const PositionalDesc *p = 0);

  /**
   * Get the value of option of type T. Throws if option is not defined.
   *
   * @param name - name of the option
   */
  template <typename T>
  T get(const String &name) const {
    try { return m_map[name].template as<T>(); }
    catch (std::exception &e) {
      HT_THROWF(Error::CONFIG_GET_ERROR, "getting value of %s: %s",
                name.c_str(), e.what());
    }
  }

  /**
   * Get the value of option of type T. Returns supplied default value if
   * not found. Try use the first form in usual cases and supply default
   * values in the config descriptions, as it validates the name and is less
   * error prone.
   *
   * @param name - name of the option
   * @param default_value - default value to return if not found
   */
  template <typename T>
  T get(const String &name, const T &default_value) const {
    try {
      Map::const_iterator it = m_map.find(name);

      if (it != m_map.end())
        return (*it).second.template as<T>();

      return default_value;
    }
    catch (std::exception &e) {
      HT_THROWF(Error::CONFIG_GET_ERROR, "getting value of %s: %s",
                name.c_str(), e.what());
    }
  }

  /**
   * Check whether a property have default value
   *
   * @param name - name of the property
   * @return true if the value is default
   */
  bool defaulted(const String &name) const {
    return m_map[name].defaulted();
  }

  bool has(const String &name) const { return m_map.count(name); }

  HT_PROPERTIES_ABBR_ACCESSORS

  /**
   * Add property to the map
   *
   * @param name - name/key of the property
   * @param v - value of the property
   * @param defaulted - whether the value is default
   */
  std::pair<Map::iterator, bool>
  add(const String &name, const boost::any &v, bool defaulted = false) {
    m_need_alias_sync = true;
    return m_map.insert(Map::value_type(name, Value(v, defaulted)));
  }

  /**
   * Set a property in the map, create if not found
   *
   * @param name - name/key of the property
   * @param v - value of the property
   * @param defaulted - whether the value is default
   */
  void set(const String &name, const boost::any &v, bool defaulted = false) {
    std::pair<Map::iterator, bool> r = add(name, v, defaulted);

    if (!r.second)
      (*r.first).second = Value(v, defaulted);
  }

  /**
   * Setup an property alias. Primary has higher priority, meaning when
   * aliases are sync'ed primary value can override secondary value
   *
   * @param primary - primary property name
   * @param secondary - secondary property name
   * @param overwrite - whether to overwrite existing alias
   */
  void alias(const String &primary, const String &secondary,
             bool overwrite = false);

  /**
   * Sync alias values. So properties that are aliases to each other
   * have the same value. Value priority: primary non-default >
   * secondary non-default > primary default > secondary default
   */
  void sync_aliases();

  /**
   * Print config key-value map
   *
   * @param out - output stream
   * @param include_default - including default values or not
   */
  void print(std::ostream &out, bool include_default = false);

  /**
   * Helper to print boost::any used by property values
   */
  static String to_str(const boost::any &);

private:
  bool m_need_alias_sync;
  Map m_map;
  AliasMap m_alias_map;
};

typedef intrusive_ptr<Properties> PropertiesPtr;

/**
 * Convenient help class for access parts of properties
 */
class SubProperties {
public:
  SubProperties(PropertiesPtr &props, const String &prefix)
    : m_props(props), m_prefix(prefix) { }

  String full_name(const String &name) const { return m_prefix + name; }

  template <typename T>
  T get(const String &name) const {
    return m_props->get<T>(full_name(name));
  }

  template <typename T>
  T get(const String &name, const T &default_value) const {
    return m_props->get(full_name(name), default_value);
  }

  bool defaulted(const String &name) const {
    return m_props->defaulted(full_name(name));
  }

  bool has(const String &name) const {
    return m_props->has(full_name(name));
  }

  HT_PROPERTIES_ABBR_ACCESSORS

private:
  PropertiesPtr m_props;
  String m_prefix;
};

} // namespace Hypertable

#endif // HYPERTABLE_PROPERTIES_H
