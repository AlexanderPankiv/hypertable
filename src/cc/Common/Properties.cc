/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"

#include <iostream>
#include <limits>
#include <fstream>
#include <string>

extern "C" {
#include <ctype.h>
#include <errno.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/stat.h>
#include <sys/types.h>
}

#include "Logger.h"
#include "Properties.h"
#include "StringExt.h"

using namespace std;
using namespace Hypertable;


void Properties::load(const char *fname) throw(std::invalid_argument) {
  struct stat statbuf;
  char *name, *value, *last, *ptr, *linep;
  std::string valstr;

  if (stat(fname, &statbuf) != 0)
    throw std::invalid_argument(string("Could not stat properties file '")
        + fname + "' - " + string(strerror(errno)));

  ifstream ifs(fname);
  string line;
  while(getline(ifs, line)) {

    linep = (char *)line.c_str();
    while (*linep && isspace(*linep))
      linep++;

    if (*linep == 0 || *linep == '#')
      continue;

    if ((name = strtok_r(linep, "=", &last)) != 0) {
      if ((value = strtok_r(0, "=", &last)) != 0) {

        ptr = name + strlen(name);
        while (ptr > name && (isspace(*(ptr-1))))
          ptr--;

        if (ptr == name)
          continue;
        *ptr = 0;

        while (*value && isspace(*value))
          value++;

        if (*value == 0)
          continue;

        ptr = value + strlen(value);
        while (ptr > value && (isspace(*(ptr-1))))
          ptr--;

        if (ptr == value)
          continue;
        *ptr = 0;

        if (*value == '\'' && *(ptr-1) == '\'') {
          value++;
          *(ptr-1) = 0;
          if (*value == 0)
            continue;
        }
        else if (*value == '"' && *(ptr-1) == '"') {
          value++;
          *(ptr-1) = 0;
          if (*value == 0)
            continue;
        }

        valstr = value;

        m_map[name] = valstr;
      }
    }
  }
}

const char *Properties::get(const char *str, const char *defaultval) {
  PropertyMap::iterator iter =  m_map.find(str);
  if (iter == m_map.end())
    return defaultval;
  return ((*iter).second.c_str());
}


/**
 *
 */
int64_t Properties::get_int64(const char *str, int64_t defaultval) {

  PropertyMap::iterator iter = m_map.find(str);
  if (iter == m_map.end())
    return defaultval;

  return parse_int_value(str, (*iter).second);
}

int Properties::get_int(const char *str, int defaultval) {
  int64_t llval = get_int64(str, defaultval);

  if (llval > numeric_limits<int>::max())
    throw std::invalid_argument(string("Integer value too large for property '")
                                + str + "'");

  return (int)llval;
}

bool Properties::get_bool(const char *str, bool defaultval) {
  PropertyMap::iterator iter = m_map.find(str);
  if (iter == m_map.end())
    return defaultval;

  if (!strcasecmp((*iter).second.c_str(), "true")
      || !strcmp((*iter).second.c_str(), "1"))
    return true;
  else if (!strcasecmp((*iter).second.c_str(), "false")
           || !strcmp((*iter).second.c_str(), "0"))
    return false;

  HT_ERRORF("Invalid value for property '%s' (%s), using defaultval value",
            str, (*iter).second.c_str());

  return defaultval;
}



std::string Properties::set(const char *key, const char *value) {
  std::string oldval;

  PropertyMap::iterator iter = m_map.find(key);
  if (iter != m_map.end())
    oldval = (*iter).second;

  m_map[key] = value;

  return oldval;
}


int Properties::set_int(const String &key, int value) {
  int64_t old_val = 0;

  if (value > numeric_limits<int>::max())
    throw std::invalid_argument(string("Integer value too large for property '")
                                + key + "'");

  PropertyMap::iterator iter = m_map.find(key);
  if (iter != m_map.end()) {
    old_val = parse_int_value(key, (*iter).second);
    if (old_val > numeric_limits<int>::max())
      throw std::invalid_argument(string("Integer value too large for property "
                                  "'") + key + "'");
  }

  m_map[key] = std::string("") + value;

  return old_val;
}


int64_t Properties::set_int64(const String &key, int64_t value) {
  int64_t old_val = 0;

  PropertyMap::iterator iter = m_map.find(key);
  if (iter != m_map.end())
    old_val = parse_int_value(key, (*iter).second);

  m_map[key] = std::string("") + (long long)value;

  return old_val;
}


int64_t Properties::parse_int_value(const String &key, const String &value) {
  const char *ptr;

  for (ptr = value.c_str(); isdigit(*ptr); ptr++)
    ;

  uint64_t factor = 1LL;
  if (*ptr != 0) {
    if (!strcasecmp(ptr, "k"))
      factor = 1000LL;
    else if (!strcasecmp(ptr, "m"))
      factor = 1000000LL;
    else if (!strcasecmp(ptr, "g"))
      factor = 1000000000LL;
    else
      throw std::invalid_argument(string("Invalid value for integer property '")
                                  + key + "' (value=" + value + ")");
  }

  string numstr = string(value.c_str(), ptr-value.c_str());

  int64_t llval = strtoll(numstr.c_str(), 0, 0);
  if (llval == 0 && errno == EINVAL)
    throw std::invalid_argument(string("Could not convert property '") + key
                                + "' (value=" + value + ") to an integer");

  return llval * factor;
}
