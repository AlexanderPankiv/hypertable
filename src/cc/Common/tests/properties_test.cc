/** -*- C++ -*-
 * Copyright (C) 2008  Luke Lu (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Hypertable. If not, see <http://www.gnu.org/licenses/>
 */

#include "Common/Compat.h"
#include "Common/Logger.h"
#include "Common/Properties.h"
#include <iostream>

using namespace Hypertable;
using namespace Property;

namespace {

void basic_test(const PropertiesDesc &desc) {
  static char *args[] = { "test", "--str", "foo", "--strs", "s1",
      "--i16", "64k", "--i32", "20M", "--i64", "8g", "--strs", "s2",
      "--i64s", "16G", "--i64s", "32G" };
  Properties props;
  props.parse_args(sizeof(args)/sizeof(char *), args, desc);
  props.print(std::cout);
}

}

int main(int argc, char *argv[]) {
  try {
    PropertiesDesc desc;
    desc.add_options()
      ("help,h", "help")
      ("boo", boo()->zero_tokens()->default_value(true), "a boolean arg")
      ("str", str(), "a string arg")
      ("strs", strs(), "a list of strings")
      ("i16", i16(), "a 16-bit integer arg")
      ("i32", i32(), "a 32-bit integer arg")
      ("i64", i64(), "a 64-bit integer arg")
      ("i64s", i64s(), "a list of 64-bit integers")
      ;

    Properties props;
    props.parse_args(argc, argv, desc);

    if (props.has("help"))
      std::cout << desc << std::endl;

    props.print(std::cout, true);
    HT_TRY("basic test", basic_test(desc));
  }
  catch (Exception &e) {
    std::cout << e << std::endl;
    return 1;
  }
  return 0;
}
