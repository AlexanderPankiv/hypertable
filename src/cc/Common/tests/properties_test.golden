boo=true (default)
i16=64000
i32=20000000
i64=8000000000
i64s=[16000000000, 32000000000]
str=foo
strs=[s1, s2]
